cumulusci/robotframework/tests/salesforce/TestListener.py

```diff
@@ -1,32 +1,84 @@
-"""
-    This hybrid library/listener can be used to verify
-    messages that have been logged.
+"""This hybrid library/listener can be used to verify messages that
+    have been logged and keywords have been called.
+
+    This works by listening for log messages and keywords via the
+    listener interface, and saving them in a cache. Keywords are
+    provided for doing assertions on called keywords and for resetting
+    the cache.
+
+    The keyword cache is reset for each test case to help keep it
+    from growing too large.
 
-    This works by listening for log messages via the listener
-    interface, and saving them in a cache. Keywords are provided
-    for searching the cache, and for resetting the cache.
 """
 from robot.libraries.BuiltIn import BuiltIn
 import re
 
 
 class TestListener(object):
+    __test__ = False  # this class should be ignored by pytest
     ROBOT_LIBRARY_SCOPE = "TEST SUITE"
     ROBOT_LISTENER_API_VERSION = 2
 
     def __init__(self):
         self.ROBOT_LIBRARY_LISTENER = self
         self.log_messages = []
+        self.keyword_cache = []
         self.capture_log = True
 
     def _log_message(self, message):
         """Called whenever a message is added to the log"""
         if self.capture_log:
             self.log_messages.append(message)
 
+    def _start_test(self, name, attrs):
+        self.reset_test_listener_keyword_cache()
+
+    def _end_keyword(self, name, attrs):
+        attrs_subset = {name: attrs[name] for name in ("status", "args")}
+        self.keyword_cache.append((name, attrs_subset))
```

Comment:
Maybe this should just be a 3-tuple of `(name, attrs["status"], attrs["args"])`?

Reply:
I'm going to leave as-is. I may want to extend the listener to include more of the attributes, and keeping them in a dictionary makes that a bit easier. Thanks for the comment though.